django/db/migrations/autodetector.py

```diff
@@ -141,6 +155,53 @@ def _detect_changes(self, convert_apps=None, graph=None):
         # Renames have to come first
         self.generate_renamed_models()
 
+        self.old_through_models = {}
+        self.new_through_models = {}
+        # Prepare a list of all m2m fields in the old and new models.
+        # This is used to detect changes to through models.
+        for app_label, model_name in sorted(set(self.old_model_keys).intersection(self.new_model_keys)):
+            old_model_name = self.renamed_models.get((app_label, model_name), model_name)
+            old_model_state = self.from_state.models[app_label, old_model_name]
+            for field_name, field in old_model_state.fields:
+                # We need the through model, even if it is auto-created and
+                # through is None in the deconstructed field.
+                old_field = self.old_apps.get_model(app_label, old_model_name)._meta.get_field(field_name)
+                if old_field.remote_field and getattr(old_field.remote_field, "through", None):
+                    # Get the through key from the rendered field.
+                    through_key = (
+                        old_field.remote_field.through._meta.app_label,
+                        old_field.remote_field.through._meta.model_name
+                    )
+                    # Get the to key from the deconstructed field.
+                    to_key = field.related_model.split('.')
+                    self.old_through_models[through_key] = {
+                        "from": (app_label, model_name),
+                        "to": to_key,
+                        "through": through_key,
+                        "field": field_name
+                    }
+            new_model_state = self.to_state.models[app_label, model_name]
+            for field_name, field in new_model_state.fields:
```

Comment:
any thought to factor this code so it's not repeated between old/new_fields versions?